hkmc2/shared/src/main/scala/hkmc2/codegen/Lifter.scala

```diff
@@ -649,6 +678,41 @@ class Lifter(handlerPaths: Opt[HandlerPaths])(using State, Raise):
       // Rewrite the rest
       val remaining = rewritten match
         
+        // We create closures once the first time we see them, then re-use them later.
+        // We store already-created closures in a set in the BlockRewriter class.
+        // This set needs to be reset after processing an if-else branch or while loop,
+        // since closures nested inside each branch may not be re-used elsewhere.
+        case Match(scrut, arms, dflt, rst) =>
+          applyPath(scrut): scrut2 =>
+            applyListOf(arms)
+              .apply:
+                case tup@(cse, blk) => k =>
+                  val blk2 = applySubBlockAndReset(blk)
+                  applyCase(cse): cse2 =>
+                    if (cse2 is cse) && (blk is blk2) then k(tup) else k(cse2 -> blk2)
+              .apply: arms2 =>
+                val dflt2 = dflt.mapConserve(applySubBlockAndReset)
+                val rst2 = applySubBlock(rst)
+                if (scrut2 is scrut) &&
+                    (arms2 is arms) &&
+                    (dflt2 is dflt) && (rst2 is rst)
+                  then b else Match(scrut2, arms2, dflt2, rst2)
+            
+        case Label(lbl, bod, rst) =>
+          val lbl2 = applyLocal(lbl)
+          val bod2 = applySubBlockAndReset(bod)
+          val rst2 = applySubBlock(rst)
+          if (lbl2 is lbl) && (bod2 is bod) && (rst2 is rst) then b else Label(lbl2, bod2, rst2)
+        case Begin(sub, rst) =>
+          val sub2 = applySubBlockAndReset(sub)
```

Comment:
Do we need to reset, here? `Begin` represents a sequence; what's defined in `sub` should be available in `rst`.

Reply:
Oh, ok